---
title: "nCounter_RNA_Analysis"
author: "Ned Cauley"
date: "8/5/24"
output: html_document
---

# This markdown and analysis based on https://github.com/bhattacharya-a-bt/CBCS_normalization/blob/master/sabry_analysis.R
# The above analysis is from the publication https://academic.oup.com/bib/article/22/3/bbaa163/5891144?login=true

```{r setup}
knitr::opts_chunk$set(echo = TRUE)
```

## Set up libraries and QC functions libraries

```{r Libraries and RUVseq functions, message=FALSE, warning=FALSE}
#library(NanoStringQCPro)
#library(NanoStringNorm)

library(ggplot2)
library(EnvStats)
library(MASS)
library(RUVSeq)
library(DESeq2)
library(limma)
library(matrixStats)
library(readxl)
library(plotly)
library(pheatmap)
library(ggrepel)
library(gridExtra)
library(grid)
library(dplyr)
library(tibble)
library(tidyr)
library(stringr)
library(ggfortify)
library(PCAtools)
library(ggpubr)

source("R/ncounter_analysis_functions.R")

# Load the annotation file
annotation_file_path <- "5_20_24_ncounter_rna_annotation_edit_NC.csv"
annotation.df <- read.csv(annotation_file_path)


# Create a new folder for the run date and results
run.folder <- "11_13_2024/all/"

results.folder <- "results/"

```

```{r Define the contrast for the DEG analysis, message=FALSE, warning=FALSE}

# Define the contrast and levels from the annotation columns for QC and DEG analysis
# Levels are: Effect, Reference. 
# This means a positive log2foldchnage = lower reference, higher effect
contrasts.list <- list(
  "CD4_count_200" = c("High", "Low"), 
  "CD4_count_175" = c("High", "Low"), 
  "CD4_count_150" = c("High", "Low"), 
  "KS_alone" = c("N", "Y"), 
  "HIV_VL" = c("Y", "N"), 
  "response" = c("Y", "N"), 
  "kshv_expression_median" = c("above", "below")
)

```


## Load the processed read counts

```{r Initial Dataframes and QC,  message=FALSE, warning=FALSE}

load(file = "ncounter_processed_rcc_counts.rda")

```

## Run RUVseq differential expression analysis

```{r, message=FALSE, warning=FALSE}

# Toggles for analysis and plots
run.de <- TRUE
export.de.results <- FALSE
export.norm.counts <- FALSE
create.volcano <- FALSE
export.volcano <- FALSE
create.heatmap <- FALSE
export.heatmap <- FALSE
export.gsea.input = FALSE
export.PCA <- FALSE


# Start a vector of HK genes to use
hk.final.list <- rcc.dfs$hk.gene.list

# Check if any HKs are DE between groups
for(contrast in names(contrasts.list)){
  
  hk.genes.checked <- check_hk_for_de(hk.genes = rcc.dfs$hk.gene.list, 
                                        raw.counts = rcc.dfs$raw.counts, 
                                        annotation = rcc.dfs$pData, 
                                        contrast.field = contrast)
  
  # Remove HK from the final list if found to be a DEG
  hk.final.list <- setdiff(hk.final.list, hk.genes.checked$hk.degs)
  
}

# Retrieve KSHV genes
kshv.genes <- rownames(rcc.dfs$raw.counts[grep("KSHV", rownames(rcc.dfs$raw.counts)), ])
kshv.genes <- c(kshv.genes, "ORF75")

# Toggle to rerun analysis
de.results.list <- list()

# Run each DE contrast
if(run.de == TRUE){
  
  for(contrast in names(contrasts.list)){
    
    contrast.levels <- contrasts.list[[contrast]]
    
    # Run DESEQ with RUVseq normalization for k sources of variability
    deseq.output <- run_RUV_DESEQ2(contrast = contrast, 
                   contrast.levels = contrast.levels, 
                   annotation.data = rcc.dfs$pData, 
                   raw.counts = rcc.dfs$raw.counts, 
                   feature.data = rcc.dfs$fData, 
                   k = 1, 
                   additional.covar = NULL, 
                   hk.genes = hk.genes.checked$hk.genes.final)
    
    de.results.list[[contrast]] <- deseq.output
    
    
    # Export all DE results
    if(export.de.results == TRUE){
      write.csv(deseq.output$all.de.results, file = paste0(results.folder, run.folder,   "DE_results_",contrast,".csv"), row.names =   FALSE)
    }
    
    # Gather normalized counts
    normalized.counts <- as.data.frame(counts(de.results.list[[contrast]]$dds, normalized = TRUE))
    # Add a gene column
    gene <- rownames(normalized.counts)
    normalized.counts <- cbind(gene, normalized.counts) 
  
  }
  
  # Export Normalized read counts
  if(export.norm.counts == TRUE){

    write.csv(normalized.counts, file = paste0(results.folder, 
    run.folder,   
    "RUV_normalized_counts.csv"), 
    row.names = FALSE)
    
  }

}
```

### Plots

#### Volcano Plot

```{r}
# Create plots per contrast

# Volcano plot
if(create.volcano == TRUE){
  
  volcano.list <- list()
  
  for(contrast in names(de.results.list)){
    
    # Standard volcano
    volcano.output <- make_volcano(de.results = de.results.list[[contrast]]$all.de.results,
                                   title = contrast, 
                                   legend.title = paste0("Expression in ", contrast.levels[1]), 
                                   x.axis.title = "log2 Fold Change", 
                                   fc.limit = 1, 
                                   custom.gene.labels = NULL, 
                                   remove.controls = TRUE, 
                                   remove.genes = kshv.genes)
    
    volcano.list[[contrast]] <- volcano.output
    
    # Export the volcano plot
    if(export.volcano == TRUE){
        ggsave(paste0(results.folder, run.folder, "volcano_", contrast, "_adjpval.png"),
             plot = volcano.output$volcano.plot, 
             width = 8, 
             height = 8, 
             units = "in")
    }
  
    
  }
  
}

```

#### Heatmap

```{r}

# Heatmap
if(create.heatmap == TRUE){
  
  heatmap.list <- list()
  
  for(contrast in names(de.results.list)){
    
    contrast.levels <- contrasts.list[[contrast]]
  
    heatmap.output <- make_heatmap(deseq.output = de.results.list[[contrast]], 
                                    contrast = contrast, 
                                    contrast.levels = contrast.levels, 
                                    annotation = annotation.df, 
                                    remove.genes = kshv.genes, 
                                    cluster.rows = TRUE, 
                                    show.row.names = TRUE, 
                                    cluster.cols = TRUE, 
                                    show.col.names = FALSE)
    
    heatmap.list[[contrast]] <- heatmap.output
    
    
    if(export.heatmap == TRUE){
    
      ggsave(paste0(results.folder, run.folder, "heatmap_", contrast, ".png"),
           plot = heatmap.output$heatmap.contrast, 
           width = 12, 
           height = 12, 
           units = "in")
      }
    
    
    
  }
  
}


```

### Preranked List for GSEA

```{r}

# Create a ranked list of genes for use in GSEA
gsea.preranked.df <- create_GSEA_preranked(counts = normalized.counts, 
                                           annotation = annotation.df, 
                                           contrast = contrast, 
                                           contrast.levels = contrast.levels)

# Export GSEA pre-ranked gene list
if(export.gsea.input == TRUE){
  write.csv(gsea.preranked.df, file = paste0(results.folder, run.folder, "GSEA/",   contrast, "_gsea_preranked_input.csv"), row.names = FALSE)
      
    }

```

# Load previously generated normalized counts

```{r}

normalized.counts <- read.csv(paste0(results.folder, 
                                     run.folder, 
                                     "RUV_normalized_counts.csv"))

# Fix column names to match annotation sheet
colnames(normalized.counts) <- gsub("\\."," ", colnames(normalized.counts))

# Add row names
rownames(normalized.counts) <- normalized.counts$gene

```

# Read count information for specific genes

```{r}

#Create a combined violin and boxplot

# Cytokine genes
cytokine.genes <- normalized.counts[grep("^IL", normalized.counts$gene), ]$gene

cytokine.genes <- c(cytokine.genes, "TNF", "IFNA1", "IFNG", "IRF4")

KS.alone.cytokine.boxplot <- gene_counts_violin_boxplot(counts = normalized.counts,
                                               annotation.df = annotation.df, 
                                               gene.list = cytokine.genes,
                                               annotation.field = "KS_alone", 
                                               display.summary.stat = FALSE, 
                                               compare.groups = TRUE)

multi.diagnosis.groups <- c("KS, KICS", 
                           "KS, MCD", 
                           "KS, MCD, PEL", 
                           "KS, PEL")

cytokine.boxplot.list <- list()

for(group in multi.diagnosis.groups){
  
  compare.groups <- c("KS", group)
  
  annotation.subset <- annotation.df %>% 
    filter(Diagnosis %in% compare.groups)
  
  group.boxplot <- gene_counts_violin_boxplot(counts = normalized.counts,
                                               annotation.df = annotation.subset, 
                                               gene.list = cytokine.genes,
                                               annotation.field = "Diagnosis", 
                                               display.summary.stat = FALSE, 
                                               compare.groups = TRUE)
  
  cytokine.boxplot.list[[group]] <- group.boxplot
  
}




```

# Boxplots of Viral Gene Probes

```{r}

# Boxplot of Lytic versus Latent for individual samples

# Identify the KSHV genes
kshv.genes <- grep("KSHV", normalized.counts$gene)
kshv.genes <- c(kshv.genes, grep("ORF75", normalized.counts$gene))

# Subset for KSHV genes
normalized.counts.kshv <- normalized.counts[kshv.genes, ]
normalized.counts.notkshv <- normalized.counts %>% 
  slice(-kshv.genes)

# Convert KSHV gene counts to log2 for plotting
normalized.log.counts.kshv <- normalized.counts.kshv %>% 
  mutate(across(where(is.numeric), log2))

# Load the KSHV gene annotation
kshv.annotation <- as.data.frame(read_excel("annotations/kshv_genes_annotation.xlsx"))

# Merge into a master DF
kshv.df <- merge(normalized.log.counts.kshv, kshv.annotation, by = "gene")


# Melt the df to a long version
kshv.df.long <- kshv.df %>%
    pivot_longer(cols = starts_with("SB"), names_to = "sample", values_to = "count")

# Gather the sample names
sample.names <- colnames(kshv.df)[grep("SB", colnames(kshv.df))]

for(ID in sample.names){
  
  # Filter for a specific sample
  kshv.df.sample <- kshv.df.long %>% filter(sample == ID)
  
  # Make a boxplot of lytic versus latent
  kshv.boxplot.type <- ggplot(kshv.df.sample, aes(x = type, y = count)) +
    geom_boxplot() +
    labs(title = paste0("Lytic vs Latent KSHV genes for ", ID),
         x = NULL,
         y = "Count")
  
  kshv.boxplot.all <- ggplot(kshv.df.sample, aes(x = sample, y = count)) +
    geom_boxplot() +
    labs(title = paste0("All KSHV gene counts for ", ID),
         x = NULL,
         y = "Count")
  
  # Combine into a single image
  ID.combined.boxplot <- grid.arrange(kshv.boxplot.all, kshv.boxplot.type, ncol = 2)
  
  # Save the plot
  export.kshv.boxplot <- FALSE
  if(export.kshv.boxplot == TRUE){
    
    ggsave(paste0(results.folder, "kshv_genes/", run.folder, ID, "_kshv_genes_boxplot.png"), 
        plot = ID.combined.boxplot, 
        width = 8, 
        height = 8, 
        units = "in")
    
  }
  
}

kshv.boxplot.all.samples <- ggplot(kshv.df.long, aes(x = sample, y = count)) +
  geom_boxplot() +
  labs(title = "Lytic vs Latent Gene Counts by Sample",
         x = "Gene Type",
         y = "Count") + 
  theme(axis.text.x = element_text(angle = 45, hjust = 1))

export.kshv.boxplot <- FALSE
  if(export.kshv.boxplot == TRUE){
    ggsave(paste0(results.folder, "kshv_genes/", run.folder, "kshv_all_boxplot.png"), 
        plot = kshv.boxplot.all.samples, 
        width = 14, 
        height = 8, 
        units = "in")
  }


# Combine sample and type
kshv.df.long$sample_type <- paste0(kshv.df.long$sample, "_", kshv.df.long$type)

# Add annotation for faceting
facet.field <- "CD4_count_200"

subset.fields <- c("SampleID", facet.field)

annotation.facet <- annotation.df %>% 
  select(all_of(subset.fields))

kshv.df.long.facet <- merge(kshv.df.long, 
                            annotation.facet, 
                            by.x = "sample", 
                            by.y = "SampleID")

# Boxplots faceted fro specific annotation groups

# By Sample

kshv.boxplot.facet.sample <- ggplot(kshv.df.long.facet, aes(x = sample, y = count, color = type)) +
  geom_boxplot() +
  labs(title = paste0("KSHV Gene Counts for ", facet.field),
         x = "Sample",
         y = "Log Counts") + 
  theme(axis.text.x = element_text(angle = 45, hjust = 1, size = 5)) + 
  facet_wrap(~ .data[[facet.field]], scales = "free_x")

export.kshv.boxplot <- FALSE
  if(export.kshv.boxplot == TRUE){
    ggsave(paste0(results.folder, run.folder, "kshv_type_boxplot_", facet.field, ".png"), 
        plot = kshv.boxplot.facet.sample, 
        width = 14, 
        height = 8, 
        units = "in")
  }


# By total annotation group

kshv.boxplot.facet.total <- ggplot(kshv.df.long.facet, aes(x = .data[[facet.field]], y = count, color = type)) +
  geom_boxplot() +
  labs(title = paste0("KSHV Gene Counts for ", facet.field),
         x = "Sample",
         y = "Log Counts") + 
  theme(axis.text.x = element_text(angle = 45, hjust = 1, size = 5)) + 
  facet_wrap(~ .data[[facet.field]], scales = "free_x")

export.kshv.boxplot <- FALSE
  if(export.kshv.boxplot == TRUE){
    ggsave(paste0(results.folder, run.folder, "kshv_type_boxplot_total_", facet.field, ".png"), 
        plot = kshv.boxplot.facet.total, 
        width = 14, 
        height = 8, 
        units = "in")
  }

# Boxplot by KSHV Gene

kshv.genes.boxplot <- ggplot(kshv.df.long, aes(x = gene, y = count, color = type)) +
  geom_boxplot() +
  labs(title = paste0("KSHV Gene Counts for all Samples"),
         x = "KSHV Gene",
         y = "Log2 Count") + 
  theme(axis.text.x = element_text(angle = 45, hjust = 1, size = 5))

export.kshv.boxplot <- FALSE
  if(export.kshv.boxplot == TRUE){
    ggsave(paste0(results.folder, "kshv_genes/", run.folder, "kshv_genes_boxplot.png"), 
        plot = kshv.genes.boxplot, 
        width = 14, 
        height = 8, 
        units = "in")
  }



export.kshv.df <- FALSE
  if(export.kshv.df == TRUE){
    write.csv(kshv.df, file = paste0(results.folder, run.folder, "kshv_gene_df.csv"), row.names = FALSE)
  }


```

# Unsupervised analysis

```{r}

# Plot PCA to check for batch effects

# Prepare the counts
rlog.dds <- rlog(de.results.list[[contrast]]$dds)

# Annotation to view PCA by
PCA.group <- "kshv_expression_median"

# PCA Plot
PCA.plot <- plotPCA(rlog.dds, intgroup = PCA.group, ntop = 700)

# Plotly PCA for interactivity
PCA.plot.data <- plotPCA(rlog.dds, intgroup = PCA.group, ntop = 700, returnData = TRUE)

PCA.plotly <- plot_ly(data = PCA.plot.data, x = ~PC1, y = ~PC2, text = ~name, mode = "markers", color = ~factor(group)) %>%
  layout(title = "PCA Plot",
         xaxis = list(title = "PC1"),
         yaxis = list(title = "PC2"))

print(PCA.plotly)

# Export the PCA plot

if(export.PCA == TRUE){
  ggsave(paste0("Results/QC/PCA.pdf"), plot = PCA.plot)
}

```

# Clustering of QC flagged samples

```{r QC Flag PCA}

qc.flagged.samples <- c("SB 22 2794", 
                        "SB 22 2963", 
                        "SB 22 696")

# Add a column for qc flagged samples
PCA.data.qc.flag <- PCA.plot.data %>% 
  mutate(qc_flag = ifelse(name %in% qc.flagged.samples, 
                          "Y", 
                          "N"))

# Genarate a plot with qc flagged samples labeled
PCA.plot <- ggplot(PCA.data.qc.flag, aes(x = PC1, 
                                         y = PC2, 
                                         color = qc_flag)) + 
  geom_point()
                     

# Plotly plot for identifying specific samples
PCA.qc.flag.plotly <- plot_ly(data = PCA.data.qc.flag, x = ~PC1, y = ~PC2, text = ~name, mode = "markers", color = ~factor(qc_flag)) %>%
  layout(title = "PCA Plot",
         xaxis = list(title = "PC1"),
         yaxis = list(title = "PC2"))

```


# Heatmap of Variable Genes

```{r}

# Prepare the normalized counts
log.norm.counts <- log2(counts(de.results.list[[contrast]]$dds, normalized = TRUE) + 1)

# Prepare the annotation columns
anno.columns <- c("KS_alone", "CD4_count_175")
annotation.df <- as.data.frame(annotation.df)

# Ensure the row names are the sample names
rownames(annotation.df) <- annotation.df$SampleID
heatmap.annotation <- annotation.df[ ,anno.columns]

# Order the annotation
heatmap.annotation <- heatmap.annotation %>% 
  arrange(CD4_count_175) %>% arrange(KS_alone)

# Define the colors for the annotations
anno.colors = list(
  KS_alone = c(Y = "seagreen", N = "red4"), 
  CD4_count_175 = c(High = "purple", Low = "coral")
  )

# Determine the most variable genes using the coefficient of variation
.calcCv <- function(x) {
  sd(x) / mean(x)
}
cv.vector <- apply(log.norm.counts, 1, .calcCv)

# Identify the NEG and POS probes
NEG.indices <- grep("NEG_", names(cv.vector))
POS.indices <- grep("POS_", names(cv.vector))
control.indices <- c(NEG.indices, POS.indices)

# Print the names of the control probes to be removed
control.probes <- names(cv.vector)[control.indices]
print(control.probes)

# Remove the control probes
cv.vector <- cv.vector[-control.indices]

# Sort the genes based on variability
cv.vector <- sort(cv.vector, decreasing = TRUE)

# Top variable genes
ngenes <- 50
goi <- names(cv.vector)[1:ngenes]

# Reorder the data columns to they match the annotation row order
row.order <- rownames(heatmap.annotation)
log.norm.counts <- log.norm.counts[, row.order]


# Top 50 genes
heatmap.top50 <- pheatmap(log.norm.counts[goi, ], 
         main = "Top 50 Variable Genes", 
         show_rownames = TRUE, 
         scale = "row", 
         show_colnames = TRUE,
         border_color = NA, 
         cluster_rows = TRUE, 
         cluster_cols = FALSE, 
         clustering_method = "average", 
         clustering_distance_rows = "correlation", 
         clustering_distance_cols = "correlation", 
         color = colorRampPalette(c("blue", "white", "red"))(120), 
         annotation_row = NA, 
         annotation_col = heatmap.annotation, 
         annotation_colors = anno.colors
         )

export.top50 <- FALSE
if(export.top50 == TRUE){
  
  ggsave("Results/heatmap_top50.png",
         plot = heatmap.top50, 
         width = 10, 
         height = 10, 
         units = "in")
  
}
```

# Clustering of Viral Gene Probes

```{r}

# See reference vignette: https://bioconductor.org/packages/release/bioc/vignettes/PCAtools/inst/doc/PCAtools.html#introduction

# Transform DESEQ normalized counts to variance stabilized
vst.dds <- varianceStabilizingTransformation(de.results.list[[contrast]]$dds)

# Gather normalized counts as a df
normalized.vst.counts <- as.data.frame(assay(vst.dds))

# Identify the KSHV genes
kshv.genes <- str_detect(rownames(normalized.vst.counts), "^KSHV")

# Subset for KSHV genes
normalized.vst.counts.kshv <- normalized.vst.counts[kshv.genes, ]
normalized.vst.counts.notkshv <- normalized.vst.counts[-kshv.genes, ]

# Transpose gene counts
normalized.vst.counts.kshv.transpose <- as.data.frame(t(normalized.vst.counts.kshv))

export.kshv.gene.counts <- FALSE
if(export.kshv.gene.counts == TRUE){
  
  write.csv(normalized.counts.kshv, file = "Results/kshv_counts_samples_as_columns.csv")
  write.csv(normalized.counts.kshv.transpose, file = "Results/kshv_counts_samples_as_rows.csv")

  
# Gather the average expression per sample
kshv.means <- colMeans(normalized.vst.counts.kshv)
notkshv.means <- colMeans(normalized.vst.counts.notkshv)

total.median.kshv <- median(kshv.means)
total.mean.kshv <- mean(kshv.means)

total.median.notkshv <- median(notkshv.means)
total.mean.notkshv <- mean(notkshv.means)

label.data <- data.frame(
  x = c(total.median.kshv, 
        total.mean.kshv, 
        total.median.notkshv, 
        total.mean.notkshv),
  y = c(0.05, 0.07, 0.59, 0.61),
  label = c(paste("KSHV Median:", round(total.median.kshv, 2)), 
            paste("KSHV Mean:", round(total.mean.kshv, 2)), 
            paste("Other Genes Median:", round(total.median.notkshv, 2)), 
            paste("Other Genes Mean:", round(total.mean.notkshv, 2)))
)

# Make a histogram of KSHV gene counts
kshv.distribution.plot <- ggplot(data.frame(kshv.means), aes(x = kshv.means)) +
  geom_histogram(aes(y = ..density..), bins = 14, fill = "lightblue", color = "black", alpha = 0.5) + 
  geom_density(alpha=0.6) +
  geom_vline(xintercept = total.median.kshv, color = "coral", linetype = "dashed", linewidth = 1) + 
  geom_vline(xintercept = total.mean.kshv, color = "maroon", linetype = "dashed", linewidth = 1) +
  geom_vline(xintercept = total.median.notkshv, color = "lightgreen", linetype = "dashed", linewidth = 1) +
  geom_vline(xintercept = total.mean.notkshv, color = "darkgreen", linetype = "dashed", linewidth = 1) + 
  labs(title = "Distribution of Average KSHV Gene Counts per Sample",
       x = "Average Gene Count",
       y = "Frequency") + 
  geom_text_repel(data = label.data, aes(x = x, y = y, label = label), color = "black", vjust = -0.5) +
  theme_minimal()

}
```

```{r}
# Visualize clustering of genes

kshv.annotation <- as.data.frame(read_excel("annotations/kshv_genes_annotation.xlsx"))

rownames(kshv_annotation) <- kshv_annotation$gene

# Ensure the rows and columns of the metadata and counts are aligned
kshv_annotation <- kshv_annotation[order(rownames(kshv_annotation)), ]
normalized.counts.kshv.transpose <- normalized.counts.kshv.transpose[ order(colnames(normalized.counts.kshv.transpose))]

# Create PCA for genes 

kshv.pca.by.gene <- pca(normalized.counts.kshv.transpose, 
                        metadata = kshv_annotation, 
                        removeVar = 0.1)


kshv.pca.plot.by.gene <- biplot(kshv.pca.by.gene, 
       colby = "type", 
       legendPosition = 'top', 
       legendLabSize = 16, 
       legendIconSize = 8.0)

export.PCA <- FALSE
if(export.PCA == TRUE){
  
  ggsave(paste0("Results/QC/kshv_pca_plot_by_sample_", contrast, ".png"),
         plot = kshv.pca.plot.by.sample, 
         width = 8, 
         height = 8, 
         units = "in")
  
  ggsave(paste0("Results/QC/kshv_pca_plot_by_gene.png"),
         plot = kshv.pca.plot.by.gene, 
         width = 8, 
         height = 8, 
         units = "in")
  
}

```

```{r Divide samples by KSHV average expression}

# Determine the average per sample
#normalized.counts.kshv <- normalized.counts.kshv %>% 
#  select(-gene)

normalized.counts.kshv <- normalized.counts.kshv %>% 
  select(-gene)
average.kshv.counts <- colMeans(normalized.counts.kshv)

# Set up a data frame for determining groups
avg.kshv.df <- data.frame(SampleID = names(average.kshv.counts), 
                          meanKSHV = unname(average.kshv.counts))

# Find the median of all samples' average KSHV gene expression
median.kshv <- median(avg.kshv.df$meanKSHV)

# Assign samples to high or low groups based on the median KSHV average
avg.kshv.df <- avg.kshv.df %>% 
  mutate(kshv_expression_median = ifelse(meanKSHV > median.kshv, "above", "below"))




# Merge with the annotation
annotation_df_cleaned <- annotation_df %>% 
  mutate(SampleID = str_replace_all(SampleID, "-", " "))

annotation.update <- merge(annotation_df_cleaned, avg.kshv.df, by = "SampleID")

# Export a new annotation with kshv group assignments
export.annotation <- FALSE
if(export.annotation == TRUE){
  
  write.csv(annotation.update,
             paste0("annotations/5_20_24_ncounter_rna_annotation_edit_NC.csv"), 
            row.names = FALSE)
  
}


```

# DSP Samples KSHV Gene Analysis

```{r}

dsp.samples <- c("SB 18 4726", "SB 22 2794", "SB 22 2963", "SB 22 696")

kshv.df.long.dsp <- kshv.df.long %>% 
  filter(sample %in% dsp.samples)


kshv.boxplot.all.dsp.samples <- ggplot(kshv.df.long.dsp, aes(x = sample, y = count)) +
  geom_boxplot() +
  labs(title = "Lytic vs Latent Gene Counts for DSP Overlap Samples",
         x = "Gene Type",
         y = "Count") + 
  theme(axis.text.x = element_text(angle = 45, hjust = 1))


kshv.boxplot.dsp.sample.types <- ggplot(kshv.df.long.dsp, aes(x = sample, y = count, color = type)) +
  geom_boxplot() +
  labs(title = "Lytic vs Latent Gene Counts for DSP Overlap Samples",
         x = "Gene Type",
         y = "Count") + 
  theme(axis.text.x = element_text(angle = 45, hjust = 1, size = 5))


kshv.genes.boxplot <- ggplot(kshv.df.long.dsp, aes(x = gene, y = count, color = type)) +
  geom_boxplot() +
  labs(title = "KSHV Gene Counts for all Samples",
         x = "KSHV Gene",
         y = "Count") + 
  theme(axis.text.x = element_text(angle = 45, hjust = 1, size = 5))

```

# DSP Samples CD4 Analysis

```{r}

CD4.counts <- normalized.counts %>% 
  filter(gene == "CD4") %>% 
  select(all_of(dsp.samples))

```


# DSP Samples DE Analysis

```{r}

# Create a subset DF with only the DSP overlap samples
dsp.samples.df <- rcc.dfs

# Subset the annotation
dsp.samples.df$pData <- dsp.samples.df$pData %>% 
  filter(SampleID %in% dsp.samples)

# Subset the counts
dsp.samples.df$raw.counts <- rcc.dfs$raw.counts %>% 
  dplyr::select(all_of(dsp.samples))

  
# Run DESEQ with RUVseq normalization for k sources of variability
deseq.output <- run_RUV_DESEQ2(contrast = "CD4_count_200", 
                               contrast.levels = c("High", "Low"), 
                               annotation.data = dsp.samples.df$pData, 
                               raw.counts = dsp.samples.df$raw.counts, 
                               feature.data = dsp.samples.df$fData, 
                               k = 1, 
                               additional.covar = NULL, 
                               hk.genes = hk.final.list)
  
  
## Make a volcano plot for the DEGs
  
# Identify the NEG and POS probes
NEG.indices <- grep("NEG_", rownames(deseq.output$all.de.results))
POS.indices <- grep("POS_", rownames(deseq.output$all.de.results))
control.indices <- c(NEG.indices, POS.indices)

# Remove the control probes
deg.results.df <- deseq.output$all.de.results[-control.indices,]

# Create a column for direction of DEGs
deg.results.df$de_direction <- "NONE"
deg.results.df$de_direction[deg.results.df$padj < 0.05 & 
                             deg.results.df$log2FoldChange > 1] <- "UP"
deg.results.df$de_direction[deg.results.df$padj < 0.05 & 
                             deg.results.df$log2FoldChange < -1] <- "DOWN"

deg.results.df$deglabel <- ifelse(deg.results.df$de_direction == "NONE", NA, deg.results.df$gene)
  
# Compute the scale for the volcano x-axis
log2.scale <- max(abs(deg.results.df$log2FoldChange))

# Establish the color scheme for the volcano plot
contrast.level.colors <- c("steelblue4", "grey", "violetred4")
names(contrast.level.colors) <- c("DOWN", "NONE", "UP")

# Make the volcano plot
volcano.plot <- ggplot(data = deg.results.df, aes(x = log2FoldChange, 
                                                  y = -log10(padj), 
                                                  col = de_direction, 
                                                  label = deglabel)) +
  geom_vline(xintercept = c(-1, 1), col = "gray", linetype = 'dashed') +
  geom_hline(yintercept = -log10(0.05), col = "gray", linetype = 'dashed') + 
  xlim(-7.5, 7.5) + 
  labs(x = "CD4 High versus Low",
       y = "-log10 adjusted p-value", 
       title = "DSP Sample Overlap DE Analysis, CD4 High versus Low") + 
  geom_point(size = 2) +
  scale_color_manual("DE Direction", 
                     values = contrast.level.colors) + 
  geom_text_repel(max.overlaps = Inf) + 
  xlim(-log2.scale-1, log2.scale+1) + 
  theme(plot.title = element_text(hjust = 0.5))
  
  # Export the volcano plot
  export.volcano <- FALSE
  if(export.volcano == TRUE){
      ggsave(paste0(results.folder, run.folder, "overlap_ncounter_volcano_plot.png"),
           plot = volcano.plot, 
           width = 8, 
           height = 8, 
           units = "in")
  }
  
  # Export Normalized read counts
  export_norm_counts <- FALSE
  if(export_norm_counts == TRUE){
    # Grab the normalized_counts
    normalized_counts <- as.data.frame(counts(deseq.output$dds, normalized = TRUE))
    
    
    # Add a gene column
    gene <- rownames(normalized_counts)
    normalized_counts <- cbind(gene, normalized_counts) 
    
    # Export
    write.csv(normalized_counts, file = paste0(results.folder, run.folder, "RUV_normalized_counts.csv"), row.names = FALSE)
    
  }
  
  # Export the DEG list
  export_sig_DEG_list <- FALSE
  if(export_sig_DEG_list == TRUE){
    write.csv(deseq.output$deg.list, file = paste0(results.folder, run.folder, "DEG_list_",contrast,".csv"), row.names =   FALSE)
  }
  
  # Export all DE results
  export_DE_results <- FALSE
  if(export_DE_results == TRUE){
    write.csv(deg.results.df, file = paste0(results.folder, run.folder, "overlap_ncounter_de_results.csv"), row.names =   FALSE)
  }
  
  


```

# CIBERSORT

```{r}

# Load the reference file
# LM22 reference: https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5895181/
lm22.file.path <- "/Users/cauleyes/Cibersortx/LM22.txt"
lm22.reference <- read.csv(lm22.file.path, sep = "\t")

# Compare probe set to CIBERSORT reference
lm22.reference.genes <- lm22.reference$Gene.symbol

# Load the normalized counts
norm.counts <- read.csv(paste0(results.folder, run.folder, "RUV_normalized_counts.csv"))

# Determine genes missing from the reference in the ncounter set
ncounter.gene.list <- norm.counts$gene

save.ref.genes <- FALSE
if(save.ref.genes == TRUE){
  save(lm22.reference.genes, 
       ncounter.gene.list, 
       file = paste0(results.folder, 
                     run.folder,
                     "cibersort_gene_lists.rda"))
}
  
missing.from.ref <- setdiff(ncounter.gene.list, lm22.reference.genes)

missing.from.ncounter <- setdiff(lm22.reference.genes, ncounter.gene.list)

found.in.ref <- intersect(ncounter.gene.list, lm22.reference.genes)


# Genes to map, from Bahman's analysis using geneSynonym
mapping.list <- c("EMR1" = "ADGRE1", 
                  "FZD3" = "FZD9", 
                  "KIR2DL1" = "KIR2DL3", 
                  "SIRPG" = "SIRPB2", 
                  "TEP1" = "PTEN")

norm.counts.mapped <- norm.counts %>%
  mutate(gene = ifelse(gene %in% mapping.list, names(mapping.list)[match(gene, mapping.list)], gene))

# Write out inout file for cibersort
export.cibersort <- FALSE
if(export.cibersort == TRUE){
  
  write.table(norm.counts.mapped,
          paste0(results.folder, 
                 run.folder, 
                 "CCBR1245_ciber_sort_input.tsv"), 
          sep = "\t")
  
}




```
